File: sortList.c
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include"ctools.h"
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"clist.h"

//#define testMod
#ifdef testMod
void test() 
{

}
#endif

#define max(a,b) (((a)>(b))?(a):(b))
#define min(a,b) (((a)<(b))?(a):(b))

#ifndef testMod
struct ListNode* sort1(struct ListNode* head, struct ListNode* nextHead)
{
    if(head==nextHead)
        return head;
    if(head->next==nextHead)
    {
        head->next=NULL;
        return head;
    }
    struct ListNode *p1=head, *p2=head;
    while(p1!=nextHead)
    {
        p1=p1->next;
        p2=p2->next;
        if(p1!=nextHead)
            p1=p1->next;
    }
    p1=sort1(head,p2);
    p2=sort1(p2,nextHead);
    struct ListNode* out=NULL,*tail=NULL;
    while(p1!=NULL&&p2!=NULL)
    {
        if(p1->val<p2->val)
        {
            if(out==NULL)
                out=p1;
            else
	            tail->next=p1;
            tail=p1;
            p1=p1->next;
            tail->next=NULL;
        }
        else
        {
            if(out==NULL)
                out=p2;
            else
                tail->next=p2;
            tail=p2;
            p2=p2->next;
            tail->next=NULL;
        }
    }
    if(p1!=NULL)
        tail->next=p1;
    else
        tail->next=p2;
    return out;
}

struct ListNode* sortList(struct ListNode* head){
    if(head==NULL||head->next==NULL)
        return head;
    head=sort1(head, NULL);
    return head;
}
#endif

int main()
{
#ifdef testMod
    test();
#endif
#ifndef testMod
    struct ListNode* head=getListNode(-1);
    head->next=getListNode(5);
    head->next->next=getListNode(3);
    head->next->next->next=getListNode(4);
    head->next->next->next->next=getListNode(0);
    struct ListNode* out=sortList(head);
    travelList(out);
#endif
    return 0;
}
    

File: sortList.java

public class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null||head.next==null) {
            return head;
        }
        ListNode p=head;
        ListNode q=head;
        ListNode prevP=head;
        while(q!=null) {
            prevP=p;
            p=p.next;
            q=q.next;
            if(q!=null) 
                q=q.next;
        }
        prevP.next=null;
        q=sortList(head);
        p=sortList(p);
        head=null;
        ListNode tail=null;
        if(p.val<q.val) {
            head=p;
            tail=p;
            p=p.next;
            tail.next=null;
        } else {
            head=q;
            tail=q;
            q=q.next;
            tail.next=null;
        }

        while(q!=null&&p!=null) {
            if(p.val<q.val) {
                tail.next=p;
                tail=p;
                p=p.next;
                tail.next=null;
            } else {
                tail.next=q;
                tail=q;
                q=q.next;
                tail.next=null;
            }
        }
        if(q!=null)
            tail.next=q;
        if(p!=null)
            tail.next=p;
        return head;
    }

    public static void main(String[] args) {
        // head = [4,2,1,3]
        ListNode a=new ListNode(4);
        ListNode b=new ListNode(2);
        ListNode c=new ListNode(1);
        ListNode d=new ListNode(3);
        a.next=b;
        b.next=c;
        c.next=d;
        ListNode.show(a);
        ListNode.show(new Solution().sortList(a));
    }
}

class ListNode {
    int val;
    ListNode next;
    public ListNode(int val) {
        this.val=val;
        next=null;
    }

    public static void show(ListNode n) {
        if(n==null)  {
            System.out.println();
            return;
        }
        System.out.print(n.val+ " ");
        show(n.next);
    }
}
